### ISSUE TITLE
Release 4.1 - March 2021

### ISSUE THREAD

**Issue by philwo** [bazelbuild/bazel#13099]
# Status of Bazel 4.1

This release will use Bazel 4.0.0 as its baseline and we will apply selected cherry-picks and backports on top of it. Please request cherry-picks that you'd like to get into Bazel 4.1.0 here via a comment.

- Expected release date: March 2021
- [List of release blockers](https://github.com/bazelbuild/bazel/labels/Release%20blocker)

To report a release-blocking bug, please file a bug using the `Release blocker` label, and cc me.

Task list:

- [x] Pick release baseline: https://github.com/bazelbuild/bazel/commit/6b33bdb1e22514304c0e35ce8e067f2175685245
- [x] Create release candidate: https://releases.bazel.build/4.1.0/rc1/
- [x] Check downstream projects: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2040
- [x] [Create draft release announcement](https://docs.google.com/document/d/1wDvulLlj4NAlPZamdlEVFORks3YXJonCjyuQMUQEmB0/edit)
- [x] Send for review the release announcement PR: https://github.com/bazelbuild/bazel-blog/pull/256
- [x] Push the release, notify package maintainers: @vbatts @petemounce @excitoon
- [x] Update the documentation
- [ ] Push the blog post
- [ ] Update the [release page](https://github.com/bazelbuild/bazel/releases/tag/4.1.0)

**Comment by philwo** [comment]
### Requested cherry picks

- [ ] https://github.com/bazelbuild/bazel/commit/34d98234324da83e93ba0d5ef5702880d5ac7c5c in https://github.com/bazelbuild/bazel/issues/12455#issuecomment-741797559
- [ ] https://github.com/bazelbuild/bazel/commit/e03cb63e059420847d6578d7cbfe93f05615c95e in https://github.com/bazelbuild/bazel/issues/12455#issuecomment-763348567
- [ ] https://github.com/bazelbuild/bazel/pull/12882 in https://github.com/bazelbuild/bazel/issues/12455#issuecomment-769415506
- [ ] Fix for https://github.com/bazelbuild/bazel/issues/12927 in https://github.com/bazelbuild/bazel/issues/12455#issuecomment-781551006

**Comment by dmivankov** [comment]
https://github.com/bazelbuild/bazel/commit/bef4bbbb47d47befe3711d06f358782ee12554f9 would be nice to have too

**Comment by gjasny** [comment]
Hello,

If it's not already fixed on the branch 24d086446f74606819dc53c3a436caa056ff05b7 /  #12879 would be nice to have in 4.1

Thanks,
Gregor

**Comment by coeuvre** [comment]
Remote Execution
- [ ] a87d7ed2411d5382bac58a20b79e09c464ad13b9
- [ ] 913a985a5c2fc3842b12c6e5f29af0fa1bccfd6a
- [ ] 5122617b8a22fee7acd86c9c48f2c2737709ca3f
- [ ] 325eb956c92530bdfda54a36a186cae4245a4f7b
- [ ] b6e3ba8abf033c6d3a318be8484021eff6a40dde
- [ ] 5b786da75837c5e29714e1d708c3cdf9a67ed32d
- [ ] 48648503729d53fdee1322fde2c8e6c05e99cff9
- [ ] bc54c648aa1f99509c7c36d5e6b570d066689209
- [ ] 92955e617b5c41713a5163dc0437c2a024b31815
- [ ] 75bd1ff8ab56d241916bde36291301fa026b2bab
- [ ] 37ee252f3744abc4511f55b5089cc52abd3ba09d
- [ ] f8d49faaa886321ea26e6d18c741368c129a37dc
- [ ] 32fc451600b6e94a015263eb1c8a63e974f6f4cc
- [ ] 97963c5bb24ac79eb3646dd61bfcf2f8a648af54
- [ ] e2b9a42a61596b0d24f0cadd6b7157b7f1efb221
- [ ] 6667ad7dd77f8d97952133052c17e7779c1430ec
- [ ] 7c081eb020186bfb16d4ef1c3832a8e946e99da1
- [ ] 17afbe4e224b359fee6415a5bd71bbedaa7843eb
- [ ] TODO: gRPC load balancing (part 5).
- [ ] TODO: Fix for #12972.
- [ ] TODO: Implementation for #13113.

External Repository
- [ ] 9804ba5b88e241ac94e25aac6288b307789f0db5
- [ ] b243584a479eb4481a9bf4f69acc899610a3b630

**Comment by EdSchouten** [comment]
I’d love it if we could get #12886 and #13085 in there. Both of those are small changes and make life easier for me...

**Comment by illicitonion** [comment]
Can I add 63bc1c7d0853dc187e4b96a490d733fb29f79664 to the External Repository list? (It's a follow-up from b243584 which is already in the list)

**Comment by philwo** [comment]
@meteorcloudy @coeuvre Should we cherry-pick the gRPC version upgrade, too?

https://github.com/bazelbuild/bazel/pull/13104
https://github.com/bazelbuild/bazel/pull/13105
https://github.com/bazelbuild/bazel/pull/13106

**Comment by philsc** [comment]
Some doc updates would be nice:
2eb1bf53d5fef13b89ee440af4f83003d1d0b50a
c71697cf33b0fbbb42fc2910bac83960edc7e855

**Comment by coeuvre** [comment]
> @meteorcloudy @coeuvre Should we cherry-pick the gRPC version upgrade, too?
> 
> #13104
> #13105
> #13106

Yes, it's the fix for #12927.

> Can I add 63bc1c7 to the External Repository list? (It's a follow-up from b243584 which is already in the list)

Sure.

**Comment by scele** [comment]
> Yes, it's the fix for #12927.

FYI, the second gRPC upgrade MR #13105 doesn't apply cleanly on 4.0 - I've rebased the commits onto 4.0 in [this branch](https://github.com/scele/bazel/commits/issue_12927_bazel4) if you want to leverage it when preparing the release branch.

**Comment by denyska** [comment]
For the external repos https://github.com/bazelbuild/bazel/commit/495ac923f398443be45c20ab29d183fe47e08911 would be nice cc @philwo @coeuvre

**Comment by coeuvre** [comment]
> For the external repos 495ac92 would be nice

Added to the list.

**Comment by illicitonion** [comment]
Also a request for dad96301d12aa77eb67399e08265a5f30f5ffd6a so that remote execution stacks can start using the newer APIs.

**Comment by liucijus** [comment]
We are waiting for https://github.com/bazelbuild/bazel/commit/ba60c0b3f9bbd00975c984244839b155e84b4c5d in rules_scala

**Comment by philsc** [comment]
dfb70ea4cae2ffffb76e9741d86c96505a6d05ad: This would let users `select()` on constraints in `filegroup()` targets.

ad241fbebd90a9f0ad65ccd0658838f57030db68: Let users use `cquery` to filter out incompatible targets.

**Comment by quval** [comment]
We'd be happy to see 1e258d2a7a5221613047e5cee0aaec5b56045d2b and d0676693310215407224c1b8e8aea9e3eddc183d get in.

**Comment by coeuvre** [comment]
> Also a request for [dad9630](https://github.com/bazelbuild/bazel/commit/dad96301d12aa77eb67399e08265a5f30f5ffd6a) so that remote execution stacks can start using the newer APIs.

Added to the list.

**Comment by philwo** [comment]
All cherry-picks that have been requested until now have been added to my or Chi's list. I'll try to harvest the cherries now and create 4.1.0rc1. :D

**Comment by philwo** [comment]
**Bazel 4.1.0rc1 is out 🚀 🌔** 

**Notes:**
- You all rock! Even though these are 53 commits spanning four months of work, they applied cleanly on top of each other starting from 4.0.0 once I got them in the right order and added a few missing ones. 😊 
- Binaries are here: https://releases.bazel.build/4.1.0/rc1/index.html
- Test results are here: https://buildkite.com/bazel/bazel-bazel/builds/15715. We'll have to fix a few failing tests in the next release candidate.
  - [platforms_test/test_platforms_can_be_overridden](https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/7295b218-ede8-4d6b-95d3-e2632152bb9e/src/test/shell/bazel/platforms_test/attempt_1.log)
  - [AnalysisTests/nonToolchainResolvingTargetsCantSelectDirectlyOnConstraints](https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/7295b218-ede8-4d6b-95d3-e2632152bb9e/src/test/java/com/google/devtools/build/lib/analysis/AnalysisTests/attempt_1.log)
  - [SkyframeTests/testRegisteredExecutionPlatforms_notExecutionPlatform](https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/7295b218-ede8-4d6b-95d3-e2632152bb9e/src/test/java/com/google/devtools/build/lib/skyframe/SkyframeTests/attempt_1.log)
  - [SkyframeTests/resolve_invalidExecutionPlatform](https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/7295b218-ede8-4d6b-95d3-e2632152bb9e/src/test/java/com/google/devtools/build/lib/skyframe/SkyframeTests/attempt_4.log)
- @scele Thank you for providing that branch! It was very helpful to get a clean cherry-pick for the gRPC upgrade.
- @katre I cherry-picked 8811e27353c2c10980faf7e4c5e44b431d2d4f1c, 27e15ad11410eb1014f5247fd0eeb31a46733c07, e87feb8ac9573cef993824f82370d0389570521d, 079bb7d69931705bb2b092c9017090e224ef3043 as dependencies of the requested 34d98234324da83e93ba0d5ef5702880d5ac7c5c. Is that OK?
- @coeuvre I cherry-picked 9a70805db543e2fb910e1c55ef3b3567362adf30 to resolve a merge conflict in the imports of `ByteStreamBuildEventArtifactUploader.java` that would otherwise occur during cherry-pick of f8d49faaa886321ea26e6d18c741368c129a37dc. Is that OK?
- @coeuvre I cherry-picked 1fcb18a0b455bfcb8e9940778f37d8c82c5ed5a0 to fix a build error caused by the `action.setPlatform` call in `RemoteSpawnRunner` which needs the updated proto file.
- @coeuvre I skipped a87d7ed2411d5382bac58a20b79e09c464ad13b9 because it was already part of 4.0.0.

**Commands ran:**
```
BASELINE_COMMIT=6b33bdb1e22514304c0e35ce8e067f2175685245 # 4.0.0
CHERRY_PICKS=(
    8811e27353c2c10980faf7e4c5e44b431d2d4f1c  # Fix error message from getPrerequisites to not print internal details. [added]
    27e15ad11410eb1014f5247fd0eeb31a46733c07  # Clean up ConfiguredTargetValueAccessor and ConfiguredTargetAccessor. [added]
    e87feb8ac9573cef993824f82370d0389570521d  # Move getConfigConditions into ConfiguredTarget. [added]
    34d98234324da83e93ba0d5ef5702880d5ac7c5c  # Change ConfiguredTargetQuery to use KeyedConfiguredTarget as a value.
    079bb7d69931705bb2b092c9017090e224ef3043  # Clean up old dependencies that are unused since 34d9823. [added]
    e03cb63e059420847d6578d7cbfe93f05615c95e  # Update bazelbuild/platforms to a current release.
    2eb1bf53d5fef13b89ee440af4f83003d1d0b50a  # Update docs and tests to use the @platforms//:incompatible constraint.
    c71697cf33b0fbbb42fc2910bac83960edc7e855  # Clarify test_suite behaviour in the Platforms docs.
    dfb70ea4cae2ffffb76e9741d86c96505a6d05ad  # Enable toolchain resolution for filegroup targets.
    24d086446f74606819dc53c3a436caa056ff05b7  # PlatformProviderUtils should ignore targets that don't have the needed provider.
    ba60c0b3f9bbd00975c984244839b155e84b4c5d  # ijar: fix manifest sections handling.
    58bb42ad7ca263a75c6eeef51482f805726663a5  # Revert "Switch to -fdebug-compilation-dir".
    bef4bbbb47d47befe3711d06f358782ee12554f9  # Update turbine. Built at google/turbine@1bbb136.
    ad241fbebd90a9f0ad65ccd0658838f57030db68  # Allow cquery to filter out incompatible targets.
    1782f0ae751569607ef88930c822ac460a1f8bb3  # Patch grpc to fix cares selecting the wrong source when building for darwin_arm64 cpu. [added]
    8f7bc2f67fafcaa8d25cfc77eaaedbf8eed2984a  # [1/3] Bump grpc to 1.33.1 to fix corruption when downloading CAS blobs.
    a573be53493c0ca9dfa45b84f855c4e1d6141fc5  # [2/3] Bump grpc to 1.33.1 to fix corruption when downloading CAS blobs.
    9b30172547f2093acb56aedf159a77d5dceffda2  # [3/3] Bump grpc to 1.33.1 to fix corruption when downloading CAS blobs
    1e258d2a7a5221613047e5cee0aaec5b56045d2b  # Allow exec groups to inherit from the rule or other exec groups.
    d0676693310215407224c1b8e8aea9e3eddc183d  # Support execution constraints per exec group.
    f1e0d346c8235c855e61afc2adb870e4b895e002  # Clean up RuleContext to use a Table instead of a Map of Maps.
    8186fbb47ab964a9affa9a0fc6315fcdbde2b5aa  # Documentation for #13110.
    321fe3b6b4e892821ee7dbf2d17dd8ae6a541913  # Prevent --repo_env from triggering unnecessary fetches.
    3ebf658cba43bbab1efc36518f0795a7d65e2d46  # Prevent a crash when using --repo_env=VAR without a value.
    913a985a5c2fc3842b12c6e5f29af0fa1bccfd6a  # Report digest of failed uploads.
    5122617b8a22fee7acd86c9c48f2c2737709ca3f  # Status error presentation with details.
    9a70805db543e2fb910e1c55ef3b3567362adf30  # Fix double shutdown of BuildEventArtifactUploader when BES+File output enabled. [added]
    325eb956c92530bdfda54a36a186cae4245a4f7b  # Add rxjava3 to third_party.
    b6e3ba8abf033c6d3a318be8484021eff6a40dde  # remote: set executable bit of an input file based on its real value.
    5b786da75837c5e29714e1d708c3cdf9a67ed32d  # Remote: correctly implement equals and hashCode.
    48648503729d53fdee1322fde2c8e6c05e99cff9  # Fixed an error that bazel binary is not executable when testing with remote execution.
    bc54c648aa1f99509c7c36d5e6b570d066689209  # Remote: Use parameters instead of thread-local storage to provide tracing metadata.
    92955e617b5c41713a5163dc0437c2a024b31815  # Remote: Use parameters instead of thread-local storage to provide tracing metadata. (Part 2)
    75bd1ff8ab56d241916bde36291301fa026b2bab  # Remote: Use parameters instead of thread-local storage to provide tracing metadata. (Part 3)
    37ee252f3744abc4511f55b5089cc52abd3ba09d  # Remote: Use parameters instead of thread-local storage to provide tracing metadata. (Part 4)
    f8d49faaa886321ea26e6d18c741368c129a37dc  # Remote: Use parameters instead of thread-local storage to provide tracing metadata. (Part 5)
    32fc451600b6e94a015263eb1c8a63e974f6f4cc  # Write/QueryWriteStatus logging refinement/addition.
    97963c5bb24ac79eb3646dd61bfcf2f8a648af54  # Remote: gRPC load balancing. (Part 1)
    e2b9a42a61596b0d24f0cadd6b7157b7f1efb221  # Remote: gRPC load balancing. (Part 2)
    6667ad7dd77f8d97952133052c17e7779c1430ec  # Remote: gRPC load balancing. (Part 3)
    7c081eb020186bfb16d4ef1c3832a8e946e99da1  # Remote: gRPC load balancing. (Part 4)
    17afbe4e224b359fee6415a5bd71bbedaa7843eb  # Implement getMessage for BulkTransferException.
    de8f69dd9ef2e06d2809a2ae3eedc97810e0d181  # Remote: gRPC load balancing. (Part 5)
    7a62c2d4e27e398f440910c81eacc384f38ca8be  # Remote: Add interoperability between Rx and ListenableFuture.
    1fcb18a0b455bfcb8e9940778f37d8c82c5ed5a0  # Update to latest remote-execution proto.
    dad96301d12aa77eb67399e08265a5f30f5ffd6a  # Set Platform on Action not just Command
    6c5a3ee0dcbb4b804f4aa85c038a378fb70eb1f9  # Remote: Add AsyncTaskCache which is used to deduplicate task executions and cache the results.
    7dc6f57cb639417c2d80cb19def22676d2b63b50  # Remote: Use AsyncTaskCache inside RemoteActionInputFetcher.
    9804ba5b88e241ac94e25aac6288b307789f0db5  # Add --experimental_repository_disable_download to allow users disable download for external repos.
    b243584a479eb4481a9bf4f69acc899610a3b630  # Report errors parsing rewriter config file.
    63bc1c7d0853dc187e4b96a490d733fb29f79664  # Downloader rewriter config has all_blocked_message.
    495ac923f398443be45c20ab29d183fe47e08911  # Allow UrlRewriter to change protocol, i.e. https->http, and http->https.
    8dbbde0037264c1db4b229a09f98a61ab4ca06b0  # Allow overriding the hostname and instance name in bytestream:// URIs.
)
git clone https://github.com/bazelbuild/bazel.git ~/bazel-release-$RELEASE_NUMBER
cd ~/bazel-release-$RELEASE_NUMBER
git remote add scele https://github.com/scele/bazel.git
git fetch scele issue_12927_bazel4
scripts/release/release.sh create $RELEASE_NUMBER $BASELINE_COMMIT "${CHERRY_PICKS[@]}"
scripts/release/release.sh push
```

**Comment by liucijus** [comment]
> We are waiting for [ba60c0b](https://github.com/bazelbuild/bazel/commit/ba60c0b3f9bbd00975c984244839b155e84b4c5d) in rules_scala

@philwo, I've tested `4.1.0rc1`. I guess to properly include [ba60c0b](https://github.com/bazelbuild/bazel/commit/ba60c0b3f9bbd00975c984244839b155e84b4c5d), bazel tools updated version needs to be included, right? Currently it seems to pick up previous stamping functionality.

**Comment by philwo** [comment]
@liucijus Indeed, sorry about that and thanks a lot for testing and letting me know!

I just asked my colleague @comius for help and he will prepare a java_tools release that we can cherry-pick into 4.1.0rc2 to ensure that we pick up the Turbine and ijar updates. 👍

**Comment by thii** [comment]
Keith already asked on the Bazel Slack but how do we get release based off the master branch now? (There’s no official answer there yet.)

Also, may I request cherry picks for commits that adds support for building Bazel targets Apple Silicon, and a release binary for it?
https://github.com/bazelbuild/bazel/commit/b39a51cf9d4b90902379f9668e3e2ed6a4a057c9
https://github.com/bazelbuild/bazel/commit/52457b18bdf5b4af8897e50e2a03486b86d3b2c6
https://github.com/bazelbuild/bazel/commit/8c7e11a81c35563d951062fe6f00e4e85b3009fb

**Comment by philwo** [comment]
@thii We will start creating the "rolling releases" that are based on the master branch in the next days. 😊 

I would personally like to have the Apple Silicon cherry picks in 4.x, but our CI infrastructure is probably not ready yet for building official release binaries. I'll see what we can do, though. (FYI @meteorcloudy)

**Comment by quval** [comment]
Thanks! :sunglasses: I saw my name twice in the list of contributors (once with the last name and once without) - is this by mistake? Haven't seen anybody else with the same name here...

**Comment by philwo** [comment]
@quval It's probably caused by two different e-mail addresses used in the Git commits. Our script that generates the contributors name list for a release is not the most sophisticated software in our repo ;) I'll manually de-dupe this for the final release ^^

**Comment by larsrc-google** [comment]
I'm a bit late to the party, but I would like the following cherrypicked:

#f8f66f3 - Make SimpleLogHandler not swallow interrupts.
#a607d9d - Never create more than one process per WorkerMultiplexer.
#80c03ef - Move sending requests and reading responses for multiplex workers into separate subthreads
#003cfcd - Allow use of JSON protocol in multiplex workers.
#308bce3 - Actively kill off still-active workers when stopping work on interrupt.
#b5d6c38 - Change short output of worker type to have the same logic as the worker creation for sandboxing vs. multiplex.
#e7a0a71 - More properly destroy workers on interrupt.
#94ac839 - Make WorkRequestHandler do a GC after some amount of CPU time has been used.
#b048282 - Use readdir for cleanExisting in WorkerExecRoot.

**Comment by brentleyjones** [comment]
Can I reuqest these two?

**Comment by brentleyjones** [comment]
Can I request these two?
https://github.com/bazelbuild/bazel/commit/f0c32de0823307ec293a882af6a9c7a48631d33b
https://github.com/bazelbuild/bazel/commit/8959dff512fe4505af786bcf2ef981ec7082a913

**Comment by ahumesky** [comment]
Hi, one more cherry pick request for the Android rules:
https://github.com/bazelbuild/bazel/commit/586de95cbd4ac0375eba539d295edcb5bb2819c6 - Updates Android remote tooling version to 0.21.0

**Comment by coeuvre** [comment]
@brentleyjones I am not confident enough to say f0c32de0823307ec293a882af6a9c7a48631d33b is production ready (need time to test the side effect). I also have some follow up PRs pending. Would you mind waiting for this change to be picked in next release?

**Comment by brentleyjones** [comment]
@coeuvre Works for me!

**Comment by philsc** [comment]
Could we also get this bug fix in?
807f2a1929e23b60b237c63fadb25af81de2e3c3

**Comment by brentleyjones** [comment]
A crash I get only in 4.1.0 and `master` (but not 4.0.0): https://github.com/bazelbuild/bazel/issues/13239

**Comment by GabrielGhe** [comment]
Can I request f4fbbd6 for `--experimental_profile_additional_tasks=local_memory_usage`

**Comment by meteorcloudy** [comment]
Please cherry-pick the following commits to add Apple Silicon support:
8c7e11a81c35563d951062fe6f00e4e85b3009fb
e96b8ca0c244e1a6a747b98b8f01e2526c9db862
ecddf60fc636844a0ac45c83e51a886687967308

Related issue: https://github.com/bazelbuild/bazel/issues/11628

**Comment by brentleyjones** [comment]
@meteorcloudy Please see https://github.com/bazelbuild/bazel/issues/13099#issuecomment-799437789

**Comment by limdor** [comment]
Any chance to cherry-pick https://github.com/bazelbuild/bazel/commit/d052ececddf54587c576f876ce13d4b8f4aacb0b
It fixes the target incompatibility for cc_import
@philsc is some other change needed apart from this one?

**Comment by philsc** [comment]
> Any chance to cherry-pick [d052ece](https://github.com/bazelbuild/bazel/commit/d052ececddf54587c576f876ce13d4b8f4aacb0b)
> It fixes the target incompatibility for cc_import
> @philsc is some other change needed apart from this one?

That would certainly be nice to cherry-pick it. It's possible that it'll require some cherry-picking that folks aren't happy with. I'm not sure.
At the very least it would require a cherry-pick of 9868455f017ddf00eaef02cb52551b45212d4136.

**Comment by comius** [comment]
> I just asked my colleague @comius for help and he will prepare a java_tools release that we can cherry-pick into 4.1.0rc2 to ensure that we pick up the Turbine and ijar updates. 👍

I released java_tools 10.6, please cherrypick:
- [ ]  d34cc2b5a622e4c9c141383e2d0671b77889c50a

**Comment by meteorcloudy** [comment]
@brentleyjones I'm adding CI support to build for Apple Silicon, see https://github.com/bazelbuild/continuous-integration/pull/1123. Although we currently don't have Apple Silicon machine on CI, we can still do cross compile for the platform.

**Comment by alexeagle** [comment]
rules_nodejs is waiting on a cherry-pick to cut our next version:

[ ] c9e2be5

**Comment by ulfjack** [comment]
It would be good to cherry-pick 9a5cd854e0613f91d52075973e2454b1e009e1ef, which fixes an issue with remote execution metadata.

**Comment by Bencodes** [comment]
It would be great to make sure the commits for https://github.com/bazelbuild/bazel/issues/13144 are cherry picked into this release.
- Turbine fix https://github.com/bazelbuild/bazel/commit/57672aca01b3be895382c952b550c9f8edf6c9f2
- `java_rules` release with the above fix https://github.com/bazelbuild/bazel/commit/c0ac45cd7a6834bc51ebd8df9df7592755739e43

**Comment by comius** [comment]
> * `java_rules` release with the above fix [c0ac45c](https://github.com/bazelbuild/bazel/commit/c0ac45cd7a6834bc51ebd8df9df7592755739e43)

java_tools 11.2 are unfortunately not compatible with Bazel 4.x.x

**Comment by philsc** [comment]
I'd also appreciate 706f5acd02363e48076dc97e37613fd968932d03 to make it in if possible.

**Comment by Bencodes** [comment]
@comius Is it possible to cut a `java_tools` release with the turbine patches?

**Comment by jlaxson** [comment]
Possible to get e3b7e17b05f13ff183a4d7efec8ec797f3f5eaa3?

**Comment by coeuvre** [comment]
Commits for 4.1.0rc2

Remote Execution
- [ ] 9a5cd854e0613f91d52075973e2454b1e009e1ef in https://github.com/bazelbuild/bazel/issues/13099#issuecomment-805364130
- [ ] 09c621e4cf5b968f4c6cdf905ab142d5961f9ddc
- [ ] 662cf54de7a103db30e04ebae2d2b919437c4846 fix for #13239
- [ ] Fix for #13262
- [ ] Implementation for #13113.

**Comment by brentleyjones** [comment]
If possible, can we get https://github.com/bazelbuild/bazel/commit/47edc57806056f3c8764241ed41b8acc72bd2ebf as well?

**Comment by philwo** [comment]
Hi folks, just wanted to let you know that I'm still on vacation this week and will review your additional fixes / cherry-picks asap when I'm back and then create 4.1.0rc2! 😊

**Comment by philwo** [comment]
Hi folks, just wanted to let you know that I'm still on vacation this week and will review your additional fixes / cherry-picks asap when I'm back and then create 4.1.0rc2! 😊

**Comment by ecunal** [comment]
Is it possible to get this fix too please https://github.com/bazelbuild/bazel/commit/f8606e5e76579442a1c6563e718ea54c673f1a04 ?

**Comment by davido** [comment]
For Java 16 (and newer) toolchain support we need this [fix](https://github.com/bazelbuild/bazel/pull/13273) for this [issue](https://github.com/bazelbuild/bazel/issues/13138), because of [JEP 396](https://openjdk.java.net/jeps/396).

It would be great if we wouldn't have to wait for years for Java 16 (and newer) toolchain support in Bazel.

**Comment by davido** [comment]
> For Java 16 (and newer) toolchain support we need this [fix](https://github.com/bazelbuild/bazel/pull/13273) for this [issue](https://github.com/bazelbuild/bazel/issues/13138), because of [JEP 396](https://openjdk.java.net/jeps/396).

^^^ aka https://github.com/bazelbuild/bazel/commit/74f35d4b2448d356b1fb86c6ba0d377f2e524aa2.

@philwo Can you cherry-pick this CL?

**Comment by illicitonion** [comment]
Can I throw in a request for 5593358a58b66f06c4e421bb48856de94c3fd625 which fixes https://github.com/bazelbuild/bazel/issues/13243?

**Comment by finn-ball** [comment]
Can we have this in please? It fixes RE + cc_test coverage:

https://github.com/bazelbuild/bazel/commit/ab1da7beb367147d7d2df17f53bbf0ed78118b69

**Comment by wisechengyi** [comment]
If possible, can we add https://github.com/bazelbuild/bazel/commit/61da1d2bf10eabba4c75de959b0374f302d89d70 as the review process took 3 months? It would be nice if we don't have wait for another release.

**Comment by daveyc123** [comment]
If it's not too late, would like to have 65d18a5221926f2beda62c1c427f8f64833109ae included as well. It just missed the 4.0 cutoff.

**Comment by konste** [comment]
@philwo Could you please pick https://github.com/bazelbuild/bazel/commit/596653d3cf76e7b208da343e1fde5fe20273a5ff for 4.1? We really need it now.

**Comment by brentleyjones** [comment]
If there is an rc3, can https://github.com/bazelbuild/bazel/commit/615e1b16a81b0defc15699ec8027d6ddd70366d1 get in it?

**Comment by philwo** [comment]
Additional cherry-picks:
- 11651824a9d0ffb9adb9611dcd39f4c95a59d750 to fix `//src/test/java/com/google/devtools/build/lib/analysis:AnalysisTests`
- 13031e5b3bd7c8f29b96b2fee1b380160e0e27fc to fix `//src/test/java/com/google/devtools/build/lib/skyframe:SkyframeTests`
- f31e86768579ad7ec57ba13f4c3c1348f5c2702e to fix `//src/test/shell/bazel:platforms_test`

**Comment by meteorcloudy** [comment]
Backporting Apple Silicon support to 4.1.0: 3974b47183bec745b46ee5fff921421970a075f8
I have squash necessary changes into one commit because they are customised for 4.1.0
I manually tested the cross-compiled binary on a M1 machine.

**Comment by illicitonion** [comment]
Apologies - a750a56f8f4061516ec3056ae8a8295ea8279903 will also need 4840a68cd273a429e46d4114a3973fd11d0e3583 as a follow-up to avoid NPEs! :)

**Comment by philwo** [comment]
@illicitonion No problem! Just in time for rc3. 😄

**Comment by philwo** [comment]
**Bazel 4.1.0rc3 is out**

**Notes:**
- Please see https://github.com/bazelbuild/bazel/issues/13099#issuecomment-785109689 for an overview of successful and failed cherry-picks.

**Changes**
```
Daniel Wagner-Hall (1):
      And mnemonic and label to remote metadata

Googler (1):
      Start the file existence check traversal from the execroot base instead of execroot so that external repo files at "<execroot>/../<path>" are correctly handled when the sibling repository layout is enabled.

John Cater (3):
      Update platforms_test to not rely on filegroup not using toolchain
      Update SkyframeTests to not rely on filegroup not using toolchain
      Update ConfigurableAttributesTest to not rely on filegroup not using …

Keith Smiley (3):
      Remove wrapped_clang params files after use
      Silence swiftmodule timestamp warnings
      Change `set -x` in coverage to be set by var

Thi Doan (1):
      Explicitly state that embedding macOS OpenJDK is for x86_64

Yun Peng (1):
      Add native support for Apple Silicon

ajurkowski (2):
      Allow using embedded tools in sandboxed spawn runners.
      Fix racy write of temporary files while staging virtual inputs for the sandbox.

chiwang (3):
      Remote: Fix a race that AsyncTaskCache#Execution could be reused after disposed which results in CancellationException("disposed") propagated to downstream.
      Remote: Use shutdownNow() instead of shutdown() in ChannelConnection#close() as a workaround to a gRPC bug.
      Remote: Check the return value of ActionOwner.getLabel() since it could be `null`.

larsrc (3):
      Make WorkerExecRoot not be a subclass of SandboxedSpawn.
      Make WorkerExecRoot not be re-created on each createFileSystem() call. Preparation for holding a map of existing links, but also just nicer.
      Use readdir for cleanExisting in WorkerExecRoot.

philwo (2):
      Fix external_path_test with newer Xcode versions.
      Bump minimal JDK install base maximum size from 290 to 295 MB.
```

**Commands ran:**
```
RELEASE_NUMBER=4.1.0
BASELINE_COMMIT=616dc264f02907d7b7887285d22307dfe6d097b6 # 4.1.0rc2
CHERRY_PICKS=(
    5f40d12e741aa30d506eaa15673fb2ae76d29468
    20f62f87c40c5f83eb65b4417cf28f1d800013b0
    eb762d4e7431637e607146b1c191485795047ef9
    f31e86768579ad7ec57ba13f4c3c1348f5c2702e
    13031e5b3bd7c8f29b96b2fee1b380160e0e27fc
    11651824a9d0ffb9adb9611dcd39f4c95a59d750
    4b68532e7ea5eb80c926b7b8e2ec2be300004628
    31db460a45767de0bcd664a6efbe9d163b85b802
    a2cc0460dc84ad2dc88019af2fe2a65ce80c61e5
    b048282c7893231d3a7191b251804973917b07a4
    3974b47183bec745b46ee5fff921421970a075f8
    9ebc65f90a71191f8f5e00653a9c52582b8fcc15
    09c621e4cf5b968f4c6cdf905ab142d5961f9ddc
    0299cd7e17203a4ce0ea947b62a7c55f1afb8225
    47edc57806056f3c8764241ed41b8acc72bd2ebf
    f6e1074b09ebefba185c0531e9cea26b9596c8a9
    a750a56f8f4061516ec3056ae8a8295ea8279903
    615e1b16a81b0defc15699ec8027d6ddd70366d1
    fe4daea99c8cd163793eca84bfb12c8fc437616b
    4840a68cd273a429e46d4114a3973fd11d0e3583
)
scripts/release/release.sh create --force_rc=3 $RELEASE_NUMBER $BASELINE_COMMIT "${CHERRY_PICKS[@]}"
git shortlog release-4.1.0rc2...release-4.1.0rc3
git diff release-4.1.0rc2...release-4.1.0rc3
bazel build //src:bazel
bazel build --build_tests_only //scripts/... //src/java_tools/... //src/test/... //src/tools/execlog/... //src/tools/singlejar/... //src/tools/workspacelog/... //third_party/ijar/... //tools/android/... //tools/aquery_differ/... //tools/python/...
scripts/release/release.sh push
```

**Comment by philwo** [comment]
Bazel 4.1.0rc3 passes all tests on CI, but I noticed that a few Android tests on Windows are broken since 4.1.0rc2. I'm running a bisect to find out if any of the cherry-picks is responsible for it, or if it's something infrastructure related.

**Comment by philwo** [comment]
This is the culprit for the hanging Android tests on Windows in 4.1.0rc2 and rc3: c4e436a76d63d373cd51e01dbc9a03eadf0d6c94.

The symptom is that the Desugar worker just hangs and never quits, which causes the test to eventually time out.

I'll investigate tomorrow. Maybe we're missing another cherry-pick, as the culprit is part of Bazel's main branch, but we don't see the test timeouts there.

**Comment by timsutton** [comment]
Would it be possible to also include `darwin-arm64` in the built releases? (Or possible by the time 4.1.0 is released?)

**Comment by UebelAndre** [comment]
I'm not quite sure how to tell, but is https://github.com/bazelbuild/bazel/commit/c2bdd034014f66ce14529cc353cda18a32320f6c included in this release?

**Comment by philwo** [comment]
@timsutton Yes, we're looking into this :)

@UebelAndre Not yet, I'll review and try to include it when I create rc4. I need to create one more rc anyway to fix the hanging worker issue on Windows.

**Comment by philwo** [comment]
Yun just merged https://github.com/bazelbuild/continuous-integration/pull/1123 so that we can build darwin-arm64 release binaries on Bazel CI. ✅

**Comment by philwo** [comment]
Potential fix for hanging worker issue was identified by @larsrc-google: 082d98772690946ed29c157e60640c97a6e1195b.

I'll prepare rc4 now.

**Comment by UebelAndre** [comment]
Did https://github.com/bazelbuild/bazel/commit/c2bdd034014f66ce14529cc353cda18a32320f6c make the cut? 😅

**Comment by philwo** [comment]
@UebelAndre I had to cherry-pick a01e94a1e7b4251670547953e93ebf2736304cf7, too, to prevent a merge conflict. Let's see how the test results are.

**Comment by philwo** [comment]
**Bazel 4.1.0rc4 is out**

**Notes:**
- Please see #13099 (comment) for an overview of successful and failed cherry-picks.
- Binaries available for download at: https://releases.bazel.build/4.1.0/rc4/index.html

**Changes**
```
Cristian Hancila (1):
      Move --repo_env to common options

ajurkowski (1):
      Allow `DiffAwareness` to share precomputed information about the workspace and propagate it to the `WorkspaceStatusAction`.

larsrc (1):
      Implement available() method for Windows subprocesses.
```

**Commands ran**
```
RELEASE_NUMBER=4.1.0
BASELINE_COMMIT=9e7e592ca51eba1a3f120320c41c845312894d74 # 4.1.0rc3
CHERRY_PICKS=(
    a01e94a1e7b4251670547953e93ebf2736304cf7
    082d98772690946ed29c157e60640c97a6e1195b
    c2bdd034014f66ce14529cc353cda18a32320f6c
)
scripts/release/release.sh create --force_rc=4 $RELEASE_NUMBER $BASELINE_COMMIT "${CHERRY_PICKS[@]}"
git shortlog release-4.1.0rc3...release-4.1.0rc4
git diff release-4.1.0rc3...release-4.1.0rc4
bazel build //src:bazel
bazel build --build_tests_only //scripts/... //src/java_tools/... //src/test/... //src/tools/execlog/... //src/tools/singlejar/... //src/tools/workspacelog/... //third_party/ijar/... //tools/android/... //tools/aquery_differ/... //tools/python/...
scripts/release/release.sh push
```

**Comment by philwo** [comment]
Folks, this is it: 4.1.0rc4 [passes the entire test suite](https://buildkite.com/bazel/bazel-bazel/builds/16085#_) and incorporates all cherry-picks that I could get in.

Please test this release thoroughly. 😊

If all looks good, we can release this in 1 - 2 weeks.

**Comment by hochhaus** [comment]
4.1.0rc4 appears to not compile upb with gcc 10 (see #12056).

According to https://github.com/bazelbuild/bazel/issues/12040#issuecomment-692585210 this fix was cherry-picked in to 3.5.1 and 3.6.0 but it seems to be missing from 4.0.0 and 4.1.0rc4. @philwo Could this be cherry-picked in?

**Comment by meteorcloudy** [comment]
@hochhaus 4.0.0 does include those cherry-picks, if you check https://github.com/bazelbuild/bazel/commit/b3ac8f60973ba60d578ae6a653cdd993a2d206d7 and https://github.com/bazelbuild/bazel/commit/26cbf776e0cdd8fbb9a40aca4c1be6f5313b6eb4, you can see those two commits are included since `3.7.0`.

**Comment by hochhaus** [comment]
Thanks @meteorcloudy. My issue is unrelated. Please ignore the noise.

**Comment by ulfjack** [comment]
It would be good to cherrypick 8ba48ad869006d87abc84e2f699fbc59eb8470f8 as a partial fix for Java coverage.

**Comment by Bencodes** [comment]
This would be another good one to get into the 4.1 release https://github.com/bazelbuild/bazel/commit/48eee8b4b447a2ad11df28dd81a2ccb65562b5f5

**Comment by konste** [comment]
Can we also cherry pick https://github.com/bazelbuild/bazel/commit/c22855857d3fb6c13ef70c48710793bbc4f195ea if possible? It is small and looks safe to take.

**Comment by philwo** [comment]
As I haven't heard any negative feedback or release blocking bugs so far, I'm planning to release Bazel 4.1.0 tomorrow or on Wednesday, depending on how things go.

@ulfjack @Bencodes @mafanasyev-tri @konste The requested cherry-picks will then be reviewed for inclusion into Bazel 4.2.0. We don't have a fixed timeline for minor releases, so we can do it relatively quickly after 4.1.0 is out. I just don't want to delay this one any longer, if it looks like a good release.

**Comment by philwo** [comment]
So, ah.. the good news: I haven't heard any negative feedback so in theory we'd be good to go! Yay.

However, I was asked to get Bazel 4.1.0 through our Google launch review process, before I may press the "push the release" button. This will need a few days, I'm sorry. I'll use the time to polish the release notes. For the next release we know that this will be needed and can just kick off the process in the beginning with rc1, so it won't be on the critical path for the release timeline.

**Comment by steeve** [comment]
> We don't have a fixed timeline for minor releases

This is a new development, is there some kind of documentation/blog as to why "one minor per month" was dropped? It looks like a fairly recent change.

Thank you for another great release!

**Comment by limdor** [comment]
> This is a new development, is there some kind of documentation/blog as to why "one minor per month" was dropped? It looks like a fairly recent change.

I guess that you are looking for this:
https://blog.bazel.build/2020/11/10/long-term-support-release.html

**Comment by philwo** [comment]
@steeve For LTS releases we don't have a fixed schedule, as the frequency of the releases depends on the availability and urgency of cherry-picks. I think monthly is still a good average to aim for, but there's nothing that would prevent us from doing more frequent releases if there's something you're all waiting for or less frequent releases if there's just nothing that needs to be cherry-picked. :)

We're also going to start publishing rolling releases really soon now, which will be cut semi-automatically ~weekly from the `master` branch and will bring you all the newest features and bug fixes. These will be based upon our internal Blaze releases: We do an internal Blaze release, then test that on the entire Google monorepo and if it proves to be stable and doesn't break anything, we'll release that same version as a Bazel rolling release. Hopefully bringing you stable Bazel releases with a higher frequency. 😊

**Comment by cpsauer** [comment]
@philwo, super excited to hear about the rolling releases coming soon! That's awesome.

**Comment by philwo** [comment]
Cross-posting from #13450:

We will unfortunately have to revert these commits in 4.1.0rc4 due to that feature being buggy and causing a regression:

- https://github.com/bazelbuild/bazel/commit/1e258d2a7a5221613047e5cee0aaec5b56045d2b (== 3486b783735531f4edf265babdc40a8730cc427b in 4.1.0rc4)
- https://github.com/bazelbuild/bazel/commit/d0676693310215407224c1b8e8aea9e3eddc183d (== cbefbd9e5453c084069219ef89403db36b591675 in 4.1.0rc4)
- https://github.com/bazelbuild/bazel/commit/f1e0d346c8235c855e61afc2adb870e4b895e002 (== 1aa544b60465765f77b714cdbf40fa082b0573a7 in 4.1.0rc4)
- https://github.com/bazelbuild/bazel/commit/8186fbb47ab964a9affa9a0fc6315fcdbde2b5aa (== 5f9456354a789436e8b5bcb03e7c0c8813afa148 in 4.1.0rc4)

**Comment by st0rm23** [comment]
@philwo can we cherry-pick this commit https://github.com/bazelbuild/bazel/commit/2d990cf2fc9607599486242c6c3d37f3d3b266ce ? I hope we can get rid of python2 dependency ASAP.

**Comment by larsrc-google** [comment]
@philwo Could you cherrypick the following:

5b95d91 - fixes occasional crash with local_lockfree_output 
779d660, f4b5e02 - fixes that dynamic execution was still waiting for workers to finish after remote winning 
b2231c5 - fixes a fallback to local execution that circumvents the strategy order
1a35dc5 - fixes occasional crash when trying to copy a non-readable output

**Comment by brentleyjones** [comment]
If another rc is being cut, can https://github.com/bazelbuild/bazel/commit/aaae8ce2881c8c1a5d4ad64f20d6e71aa372cf2e make it in? It helps Apple Silicon support.

**Comment by wrowe** [comment]
I don't yet see a 4.1.0 release, but also don't see a 4.2.0 fix issue so I'll bring this here;

https://github.com/bazelbuild/bazel/commit/6dc941e58dfc1d4a9714a76b921fbe11fce658ed fixes a critical issue for windows software releases from GCP RBE, a bandaid to preserve any release-build symbolic debugging information. If this can either hit an rc of 4.1.0 or if a 4.2.x ticket is opened, it would be helpful to consider this asap, I have no practical means to test if it isn't on the 4.x branch (experimenting with current master is a little ahead of the project's progress.)

**Comment by robfig** [comment]
https://github.com/bazelbuild/bazel/commit/c9f9eedc313fc043755dd30ae7e01478eed7c417 is an important fix to support static analysis tooling with rules_go. The original bug was reported over a year ago, so would love to get this in 4.1. Thank you!

**Comment by fmeum** [comment]
https://github.com/bazelbuild/bazel/commit/7acf9ea3fea4850d21114d84fbc5b663d105d3c7 causes build failures explained https://github.com/bazelbuild/bazel/issues/13464, could it be reverted for 4.1.0?

**Comment by philwo** [comment]
**Bazel 4.1.0rc5 with the reverted problematic commits has been created:**

- All tests pass: https://buildkite.com/bazel/bazel-bazel/builds/16313
- Please see https://github.com/bazelbuild/bazel/issues/13099#issuecomment-836744453 for the list of cherry-picks that had to be reverted in this release candidate.
- Binaries available for download at: https://releases.bazel.build/4.1.0/rc5/index.html

I would like to release this beginning of next week. @glukasiknuro could you please verify that this fixes the issue you saw with 4.1.0rc4?

I'll see that we get the remaining requests into 4.1.1 or 4.2.0 as soon as possible. I acknowledge that they look important, but I wanted to get 4.1.0 out, at last, without introducing more risks.

@fmeum The culprit you identified is already part of Bazel 4.0.0, thus we cannot revert it for 4.1.0. A fix or revert of that commit has to be created and pushed to the `master` branch first. If it doesn't cause issues, we can include the cherry-pick of that revert into the next 4.x release then!

**Commands ran**

```
git checkout -b release-4.1.0rc4-patches release-4.1.0rc4
git revert --no-edit --signoff \
    5f9456354a789436e8b5bcb03e7c0c8813afa148 \
    1aa544b60465765f77b714cdbf40fa082b0573a7 \
    cbefbd9e5453c084069219ef89403db36b591675 \
    3486b783735531f4edf265babdc40a8730cc427b

RELEASE_NUMBER=4.1.0
BASELINE_COMMIT=d7371772a884a7e630eb158c0e2cb4f7f8c143db # 4.1.0rc4
CHERRY_PICKS=(
    0e00b2703b79bd435d42536aed4bfcb8a2f75e6c  # Revert "Documentation for #13110"
    45cffa9b5e11fb32a97a51cd8f733434cdaf18af  # Revert "Clean up RuleContext to use a Table instead of a Map of Maps."
    9ba6469e3164c444bc2e534f4c8590049cddd3e4  # Revert "Support execution constraints per exec group"
    3ead920f83b5a0061682ca33d3d20428bf3852fb  # Revert "Allow exec groups to inherit from the rule or other exec groups."
)
scripts/release/release.sh create --force_rc=5 $RELEASE_NUMBER $BASELINE_COMMIT "${CHERRY_PICKS[@]}"
git shortlog release-4.1.0rc4...release-4.1.0rc5
git diff release-4.1.0rc4...release-4.1.0rc5
bazel clean --expunge
bazel build //src:bazel
bazel build --build_tests_only //scripts/... //src/java_tools/... //src/test/... //src/tools/execlog/... //src/tools/singlejar/... //src/tools/workspacelog/... //third_party/ijar/... //tools/android/... //tools/aquery_differ/... //tools/python/...
scripts/release/release.sh push
```

**Comment by glukasiknuro** [comment]
@philwo confirming that the issue with exec_compatible_with is fixed in rc5 - both in reproduction in https://github.com/bazelbuild/bazel/issues/13450 and actual codebase where the problem was detected. Thank you!

**Comment by philwo** [comment]
Bazel 4.1.0 has been released. Thank you all for your patience and help with getting this out the door.

Cherry-picks that could not be included will be considered for 4.1.1 or 4.2.0 asap.

**Comment by philwo** [comment]
Hey folks,

we noticed that bazel-toolchains has a broken version check and Bazel 4.1.0 triggers a code path in it that will cause all builds using bazel-toolchains to fail. This is not a bug in Bazel and there's nothing we could have done to prevent this, apart from never bumping our version to 4.1.0 in the first place. :/

The fix is to upgrade to bazel-toolchains 5.1.0, which has a fixed version check. Because bazel-toolchains has changed in a major way with version 5.0.0, this might not be straightforward (but still recommended). We are looking into providing a simple patch file that would fix the version check for bazel-toolchains 4.0.0 and which might be simpler to apply if you can't upgrade yet to 5.0.0.

**Comment by philwo** [comment]
The patch for bazel-toolchains 4.0.0 is available here:

https://github.com/bazelbuild/bazel/blob/5b4320ec65f8da791354f9cc058f3ec2e2107cfb/third_party/bazel-toolchains/bazel-toolchains-4.0.0-fix-version-check.patch

You can use it in your http_archive rule in the WORKSPACE file via `patches` and `patch_args` like this:

https://github.com/bazelbuild/bazel/commit/f8f9dc8682335b43b707a7e798262eb1d13b9eb4#diff-a7b969a66e4cf15e076089c4b3ddca51a3e02750d07938fbb7607af2645fdb9e

> Could a 4.0.x release be made?

Done: https://github.com/bazelbuild/bazel-toolchains/releases/tag/4.1.0

**Comment by msiebuhr** [comment]
I'd like to ask for the 4.1.0 (or indeed all versions) to have docker images available at <https://l.gcr.io/google/bazel>? I expect this is  sort-of Google-internal thing, but I suspect a few Googles to be involved here :)

Alternatively, the documentation at https://docs.bazel.build/versions/4.0.0/bazel-container.html should be updated to point at an up-to-date registry, as `:latest` on l.gcr.io/google/bazel refers to Bazel 3.5.0.

**Comment by philwo** [comment]
@msiebuhr We will provide Docker containers for Bazel versions shortly. 👍

**Comment by vbatts** [comment]
sorry for the delay. Fedora and Centos builds done. 
Fedora 34 and rawhide(x86_64 and aarch64) builds failed. Looking like another issue with latest gcc.
https://copr.fedorainfracloud.org/coprs/vbatts/bazel/build/2215183/

```
[1AERROR: /builddir/build/BUILD/bazel-4.1.0/third_party/ijar/BUILD:47:11: Compiling third_party/ijar/zlib_client.cc failed: (Exit 1): gcc failed: error executing command 
  (cd /var/tmp/bazel_TIraKP60/out/execroot/io_bazel && \
  exec env - \
    PATH=/builddir/build/BUILD/bazel-4.1.0/bin-hack:/builddir/.local/bin:/builddir/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/sbin \
    PWD=/proc/self/cwd \
  /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x' -MD -MF bazel-out/k8-opt/bin/third_party/ijar/_objs/zlib_client/zlib_client.d '-frandom-seed=bazel-out/k8-opt/bin/third_party/ijar/_objs/zlib_client/zlib_client.o' -iquote . -iquote bazel-out/k8-opt/bin -isystem third_party/zlib -isystem bazel-out/k8-opt/bin/third_party/zlib -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c third_party/ijar/zlib_client.cc -o bazel-out/k8-opt/bin/third_party/ijar/_objs/zlib_client/zlib_client.o)
Execution platform: //:default_host_platform
[401 / 995] Compiling src/main/cpp/blaze.cc; 4s local

[1AIn file included from third_party/ijar/zlib_client.cc:20:
./third_party/ijar/zlib_client.h:64:46: error: 'numeric_limits' is not a member of 'std'
   64 |   static const size_t MAX_BUFFER_SIZE = std::numeric_limits<int32_t>::max();
      |                                              ^~~~~~~~~~~~~~
./third_party/ijar/zlib_client.h:64:68: error: expected primary-expression before '>' token
   64 |   static const size_t MAX_BUFFER_SIZE = std::numeric_limits<int32_t>::max();
      |                                                                    ^
./third_party/ijar/zlib_client.h:64:71: error: '::max' has not been declared; did you mean 'std::max'?
   64 |   static const size_t MAX_BUFFER_SIZE = std::numeric_limits<int32_t>::max();
      |                                                                       ^~~
      |                                                                       std::max
In file included from /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/algorithm:62,
                 from third_party/ijar/zlib_client.cc:16:
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/stl_algo.h:3467:5: note: 'std::max' declared here
 3467 |     max(initializer_list<_Tp> __l, _Compare __comp)
      |     ^~~
```
https://download.copr.fedorainfracloud.org/results/vbatts/bazel/fedora-34-x86_64/02215183-bazel4/build.log.gz

**Comment by davido** [comment]
@vbatts This is a known [issue](https://github.com/bazelbuild/bazel/issues/12702) with gcc 11: `gRPC` and `absl_cpp` must be bumped.

**Comment by vbatts** [comment]
Thanks. Is that coming in a next release, or something that I need to do as
a packager?
